[Lens] Supports multi rows headers for the table visualization

stratoula opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/122965

It adds an extra setting for enabling multi rows for the table header.
The user can select among:
- single option (only 1 row)
- auto (show everything)
- custom (user sets the row lines)

It defaults to the single setting.
![lens_multi_rows_header](https://user-images.githubusercontent.com/17003240/158123459-6bb73eb2-3145-457e-aac7-26748ea61ed9.gif)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

stratoula commented:
@MichaelMarcialis as you will see in the description, I added another set of settings for the datatable that concern the header rows. 

The `lines per row` setting appear when you hit custom (same as the row height). I am not very satisfied with the result . Do you have any idea on how we can group them better? Or maybe if we change the labels somehow? 🤔 

<img width="392" alt="image" src="https://user-images.githubusercontent.com/17003240/157689799-fe7c6235-c6fc-46c5-89d2-32f871dc4796.png">

flash1293 commented:
FWIW I think it's an important case to have multi-line header **without** having multi line cells. There are quite some cases with long column names but small numeric values in the cells.

flash1293 commented on x-pack/plugins/lens/public/datatable_visualization/components/columns.tsx:
```diff
@@ -211,10 +213,24 @@ export const createGridColumns = (
+      headerRowHeight === 'auto' || headerRowHeight === 'custom'
+        ? {
+            whiteSpace: 'normal' as const,
+            display: '-webkit-box',
```
If it's totally necessary this is fine, but if we can let's use emotion for this so we don't have to do prefixing ourselves.

flash1293 reviewed (commented):
It seems like specific line count is not respected if there is no header with that height which is different than the data row height behavior:
<img width="868" alt="Screenshot 2022-03-10 at 16 17 58" src="https://user-images.githubusercontent.com/1508364/157696199-b70b60e2-d3fd-4ad3-9ef3-b6e28692b6c8.png">

I see how this would be hard to fix, I don't think it's a blocker though, just something to point out.

LGTM otherwise, but I agree we can improve the look of the UI

MichaelMarcialis commented:
> @MichaelMarcialis as you will see in the description, I added another set of settings for the datatable that concern the header rows.
> 
> The `lines per row` setting appear when you hit custom (same as the row height). I am not very satisfied with the result . Do you have any idea on how we can group them better? Or maybe if we change the labels somehow? 🤔
> 
> <img alt="image" width="392" src="https://user-images.githubusercontent.com/17003240/157689799-fe7c6235-c6fc-46c5-89d2-32f871dc4796.png">

Hey, @stratoula! Good question. My first instinct would be to make the following changes:

- Change the `Row height` label to be `Cell row height`.
- Change the `Header height` label to be `Header row height`.
- Rather than having the range sliders for header and cell row height appear within their own form row, can we migrate them to be within the header and cell row height form rows, but still positioned below their respective button group? The idea is to have the range sliders appear in the same position that you have them in currently, but as they would no longer be in their own form row, they wouldn't have their own separate label. Instead, they would appear to be an inset/nested part of the preceding button group.
- Make the new header row height group (both button group and range slider) appear before the cell row height group.

Let me know if that makes sense. Thanks!

stratoula commented:
I love it!

stratoula commented:
@elasticmachine merge upstream

stratoula commented:
To wrap up:
- I implemented @MichaelMarcialis 's suggestion (look much better now!)
- I used emotion
- I added a max of 5 lines per row for the header settings

stratoula commented:
@elasticmachine merge upstream

dej611 commented on x-pack/plugins/lens/public/datatable_visualization/components/toolbar.tsx:
```diff
@@ -49,6 +40,29 @@ export function DataTableToolbar(props: VisualizationToolbarProps<DatatableVisua
     [state, setState]
   );
 
+  const onChangeHeaderHeight = useCallback(
```
nit: it would be nice to refactor these two functions with those above, but it is not urgent.

dej611 reviewed (approved):
Tested locally with Safari and it works as expected.

The only bug I found is this edge case where the last two chars falls below the menu arrow:
<img width="964" alt="Screenshot 2022-03-16 at 13 03 36" src="https://user-images.githubusercontent.com/924948/158600963-f984197a-97df-4d18-985e-e5a9ebe7af8a.png">

stratoula commented:
@elasticmachine merge upstream
